File: backend/models/user_receipt.py
from sqlalchemy import Numeric, text
from sqlalchemy.dialects.postgresql import JSONB

from models import db


class UserReceipt(db.Model):
    __tablename__ = "user_receipts"

    id = db.Column(
        db.BigInteger, primary_key=True
    )  # Using BigInteger for better scalability
    user_id = db.Column(db.BigInteger, db.ForeignKey("users.id"), nullable=True)
    image_path = db.Column(
        db.Text, nullable=True
    )  # Using Text for unlimited length URLs
    created_at = db.Column(
        db.TIMESTAMP(timezone=True), server_default=text("CURRENT_TIMESTAMP")
    )

    # Denormalized fields extracted from receipt_data (RegularReceipt / TransportationTicket)
    is_receipt = db.Column(db.Boolean, nullable=True, default=True)
    document_type = db.Column(db.Text, nullable=True)

    # Regular receipt fields
    merchant = db.Column(db.Text, nullable=True)  # Using Text for unlimited length
    date = db.Column(db.Date, nullable=True)  # Using proper Date type
    subtotal = db.Column(Numeric(12, 2), nullable=True, default=0)
    tax = db.Column(Numeric(12, 2), nullable=True, default=0)
    tip = db.Column(Numeric(12, 2), nullable=True, default=0)
    gratuity = db.Column(Numeric(12, 2), nullable=True, default=0)
    total = db.Column(Numeric(12, 2), nullable=True, default=0)
    payment_method = db.Column(db.Text, nullable=True)
    tax_included_in_items = db.Column(db.Boolean, nullable=True, default=False)
    display_subtotal = db.Column(Numeric(12, 2), nullable=True, default=0)
    items_total = db.Column(Numeric(12, 2), nullable=True, default=0)
    pretax_total = db.Column(Numeric(12, 2), nullable=True, default=0)
    posttax_total = db.Column(Numeric(12, 2), nullable=True, default=0)
    final_total = db.Column(Numeric(12, 2), nullable=True, default=0)

    # Transportation ticket-specific fields
    carrier = db.Column(db.Text, nullable=True)  # Using Text for unlimited length
    ticket_number = db.Column(db.Text, nullable=True)  # Using Text for unlimited length
    origin = db.Column(db.Text, nullable=True)  # Using Text for unlimited length
    destination = db.Column(db.Text, nullable=True)  # Using Text for unlimited length
    passenger = db.Column(db.Text, nullable=True)  # Using Text for unlimited length
    class_ = db.Column("class", db.Text, nullable=True)
    fare = db.Column(Numeric(12, 2), nullable=True, default=0)
    currency = db.Column(db.Text, nullable=True)
    taxes = db.Column(Numeric(12, 2), nullable=True, default=0)

    # Additional metadata field for flexible data storage
    receipt_metadata = db.Column(
        JSONB, nullable=True, default=dict
    )  # Using JSONB for better performance and querying

    user = db.relationship("User", backref=db.backref("receipts", lazy=True))

    # Relationship to normalized line items
    line_items = db.relationship(
        "ReceiptLineItem",
        backref="receipt",
        lazy=True,
        cascade="all, delete-orphan",
        order_by="ReceiptLineItem.id",
    )

    def __repr__(self):
        return f"<UserReceipt {self.id}>"

File: frontend/src/zero/schemas/user-receipt.ts
/**
 * User receipt schema for Zero.
 *
 * Represents a receipt (or document) owned by a user. Stores denormalized
 * receipt data, including regular receipts and transportation tickets.
 *
 * Database table: user_receipts
 * - Denormalized fields from backend user_receipt model (merchant, totals, etc.).
 * - image_visibility: 'public' | 'owner_only'.
 * - Optional user_id links to users; supports receipt_metadata JSONB.
 *
 * Relationships:
 * - Many-to-one with users (user_receipts.user_id -> users.id)
 * - One-to-many with receipt_line_items (user_receipts.id -> receipt_line_items.receipt_id)
 */

import {
  boolean,
  json,
  number,
  string,
  table,
} from '@rocicorp/zero';

export const userReceipt = table('user_receipts')
  .columns({
    id: number(),
    user_id: number().optional(),
    image_path: string().optional(),
    image_visibility: string().optional(), // 'public' | 'owner_only'
    created_at: number(),
    is_receipt: boolean().optional(),
    document_type: string().optional(),
    merchant: string().optional(),
    date: number().optional(),
    subtotal: number().optional(),
    tax: number().optional(),
    tip: number().optional(),
    gratuity: number().optional(),
    total: number().optional(),
    payment_method: string().optional(),
    tax_included_in_items: boolean().optional(),
    display_subtotal: number().optional(),
    items_total: number().optional(),
    pretax_total: number().optional(),
    posttax_total: number().optional(),
    final_total: number().optional(),
    carrier: string().optional(),
    ticket_number: string().optional(),
    origin: string().optional(),
    destination: string().optional(),
    passenger: string().optional(),
    class_: string().from('class').optional(),
    fare: number().optional(),
    currency: string().optional(),
    taxes: number().optional(),
    receipt_metadata: json().optional(),
  })
  .primaryKey('id');
